Add EURUSD, CHFUSD and GBPUSD as price identifiers (#102)
UMAprotocol/UMIPs: UMIPs/umip-29.md

## Headers
| UMIP-29    |                                                          |
|------------|----------------------------------------------------------|

> | UMIP Title | Add EURUSD, CHFUSD and GBPUSD                            |

Reviewer: Naming convention looks good!

## Motivation
The DVM currently does not support the EURUSD, GBPUSD and CHFUSD index.

### Cost: 
While Forex data are free through open centralized exchange, brokers, and other sources APIs, the most reliable are paid. We propose to use TraderMade's APIs whose pricing is accessible on their website. TraderMade has also a Free Tier which can be used by the voters and would be sufficient to provide the price of a certain asset.

### Opportunity: 
Synthetic tokens that track Forex pairs such as EURUSD, GBPUSD and CHFUSD could be used both for speculations and hedging, but we see a bigger opportunity for using them as building blocks for helping other DeFi protocols and dApp addressing the european market.

## Technical Specification

The definition of this identifiers should be:

-----------------------------------------
- Identifier name: **EURUSD**
- Base Currency: EUR
- Quote Currency: USD

-----------------------------------------

- Identifier name: **GBPUSD**
- Base Currency: GBP
- Quote Currency: USD

-----------------------------------------

- Identifier name: **CHFUSD**
- Base Currency: CHF
- Quote Currency: USD

-----------------------------------------

> Source: https://marketdata.tradermade.com <br>

Reviewer: Thanks for fixing the formatting

> Result Processing: Median (if more than 1 source is utilized), otherwise Exact Price <br>

Reviewer: I think Median is slightly confusing with two sources. Also, this should really be for the recommended way of querying the price. I think providing the alternative sources as options for voters is more than enough. I would recommend removing median.

```suggestion
Result Processing: Exact Price <br>
```

Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus. <br>
Price Steps: 0.00001 (5 decimals in more general trading format) <br>
Rounding: Closest, 0.5 up (> .000005 rounds up, < .000005 rounds down) <br>

> Decimal: 18 (1e18) <br>

Reviewer: This decimal specification should be removed as decimals within the UMIP specify rounding decimals not scaling decimals and the identifier will be treated as 18 decimals of scaling by default

> Decimal: 18 (1e18) <br>

Reviewer: Let's call this Scaling Decimals to differentiate from the price steps

Author: Fixed and pushed :)

Pricing Interval: 60 seconds <br>
Dispute timestamp rounding: down <br>

## Rationale
Apart from the weekend, there is little to no difference in prices on liquid major Forex pairs like EURUSD, so any price feed could be used; however, for convenience, we recommend using the one of TraderMade.

## Implementation

> The value of this identifier for a given timestamp should be determined by querying for the price of EURUSD, GBPUSD and CHFUSD from TraderMade.com for that timestamp. More specifically, users can query “https://marketdata.tradermade.com/api/v1/minute_historical?currency=EURUSD&date_time=2020-11-11-13:01&api_key=apikey” and use the close price as reference. 

Reviewer: This should really be the price request timestamp rounded back to the nearest minute, correct?

Author: I believe so yes @smb2796

### USDCHF value

As most of the price feed does not provide a price for CHFUSD but USDCHF, the value of this identifier will undergo one additional step: CHFUSD = 1/USDCDF.  Token holders should take care to confirm that the order of the quote and base currency they refer to matches the one being requested by the DVM in the event of a price request.
TraderMade is provided as an accessible source to query for this data, but ultimately how one queries for these rates should be varied and determined by the voter to ensure that there is no central point of failure.
While it's important for tokenholders to have redundancy in their sources, bots and users that interact with the system in realtime need fast sources of price information. In these cases, it can be assumed that the exchange median is accurate enough

### Price feed - liquidations and disputes

Liquidation and dispute bots should have their own subscription to price feeds. Our price-feed provider’s API documentation can be found [here](https://marketdata.tradermade.com/documentation).

> In the case of a TraderMade outage voters can turn to any other available price feed API or a broker API, as the price feeds for the forementioned financial assets does not differ much between different providers. There might be some slight differences, however they are quite insignificant and would not affect the liquidation or dispute processes. In that case we'll use a median price from the different price feeds provided by the voters.

Reviewer: ```suggestion
In the case of a TraderMade outage voters can turn to any other available price feed API or a broker API, as the price feeds for the forementioned financial assets does not differ much between different providers. There might be some slight differences, however they are quite insignificant and would not affect the liquidation or dispute processes. For this case, we provide options for additional price feed providers that voters could utilize.
```

### Opportunity: 
Synthetic tokens that track Forex pairs such as EUR/USD, GBP/USD, CHF/USD or commodities such as XAU/USD could be used both for speculations and hedging, but we see a bigger opportunity for using them as building blocks for helping other DeFi protocols and dApp addressing the european market.

## Technical Specification

The definition of this identifiers should be:

-----------------------------------------
- Identifier name: **EUR/USD**
- Base Currency: EUR
- Quote Currency: USD
- Decimals: 18

-----------------------------------------

- Identifier name: **GBP/USD**
- Base Currency: GBP
- Quote Currency: USD
- Decimals: 18

-----------------------------------------

- Identifier name: **CHF/USD**
- Base Currency: CHF
- Quote Currency: USD
- Decimals: 18

-----------------------------------------

- Identifier name: **XAU/USD**
- Base Currency: XAU
- Quote Currency: USD
- Decimals: 18

-----------------------------------------

> Source: https://marketdata.tradermade.com

Reviewer: Formatting here is not showing up correctly

Author: @smb2796 Can you give me more details as I am not seeing an issue here locally?

Reviewer: Looks like this: 
![image](https://user-images.githubusercontent.com/44212783/103354870-b484ce00-4a7a-11eb-9ad1-992df45be260.png)

Result Processing: Median (if more than 1 source is utilized), otherwise Exact Price
Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.
Price Steps: 0.0001 (4 decimals in more general trading format)
Rounding: Closest, 0.0001 up

> Decimal: 18 (1e18)

Reviewer: To be clear, you want prices to be rounded to 4 decimal places, but still have prices returned with 18 decimal places?

Author: Yes, that is correct. The idea is that 18 decimals are set when writing on the blockchain otherwise 4 decimals of rounding are used. As ex. EUR/USD price is 1.23456 which rounds to 1.2345, but on the blockchain it is displayed 12345000000000000000000. @smb2796

Pricing Interval: 60 seconds
Dispute timestamp rounding: down

## Rationale
Apart from the weekend, there is little to no difference in prices on liquid major Forex pairs like EURUSD, so any price feed could be used; however, for convenience, we recommend using the one of TraderMade.

## Implementation

> The value of this identifier for a given timestamp should be determined by querying for the price of EURUSD, GBPUSD and CHFUSD from TraderMade.com for that timestamp. More specifically, users can query “https://marketdata.tradermade.com/api/v1/minute_historical?currency=EURUSD&date_time=2020-11-11-13:01&api_key=apikey” and use the close price as reference. 

Reviewer: Need XAU/USD here as well

Author: Thank you for the pointer :)

Author: Found a few more places actually. Pushing a commit :) @smb2796

Reviewer: As discussed offline, the existing XAUUSD identifier should suffice

Tradermade’s price feed is an aggregated feed from multiple Tier One and Two Banks, Market-Makers and Data Providers. They are popular with Quantitative Traders, Fintech companies and Institutional customers who require a high quality and trusted feed.

### Weekend timestamp

Over the weekend the REST API does not return any price, but we can request the price of a certain moment before the market close (as ex: the closing price of Friday).

> Due to unavailability of price feed for any financial asset over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which esentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset was created, if it was created over the weekend. If not the closing price on Friday for a certain asset should be a navigating point in calculating the collateralization ratio of a position and in the liquidation process.

Reviewer: Small nitpick, but more specifically a lack of availability for FX here. 

```suggestion
Due to unavailability of price feed for foreign exchange rates over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which essentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset was created, if it was created over the weekend. If not the closing price on Friday for a certain asset should be a navigating point in calculating the collateralization ratio of a position and in the liquidation process.
```

- Historical data accessible with the Basic Paid Plan - 10$ per month

Example requests:

- Live Rate for EURUSD: `https://fcsapi.com/api-v3/forex/latest?symbol=EUR/USD&access_key=api_key`
- Historical price for a certain date for EUR/USD - gives back the OHLC: `https://fcsapi.com/api-v3/forex/history?symbol=EUR/USD&period=1D&from=2020-12-23&to=2020-12-24&access_key=api_key`
- Price of EUR/USD between certain dates and times - gives a breakdown on the price each 1 minute between 12:00 and 12:30: `https://fcsapi.com/api-v3/forex/history?symbol=EUR/USD&period=1M&from=2020-12-23T12:00&to=2020-12-23T12:30&access_key=api_key`  

## Security considerations
Adding this new identifier by itself poses a little security risk to the DVM or priceless financial contract users. However, anyone deploying a new priceless token contract referencing this identifier should take care to parameterize the contract appropriately to avoid the loss of funds for synthetic token holders. Additionally, the contract deployer should ensure that there is a network of liquidators and disputers ready to perform the services necessary to keep the contract solvent.

> While the price of EURUSD does not vary much across various price feed during the weekday, it may be more challenging to get a unified price feed on the weekend: first, while the underlying market is not closed, most of the broker and banks feeding the APIs price feed are; then, for the one providing a price feed, they often add a very big spread, which makes the price of forex pairs vary significantly by source; a way to mitigate it is to use the last price known. 

Reviewer: Small nit but doesn't this apply to all of these rates, not just EURUSD?

Author: Fixed it :)

## Headers
| UMIP-29    |                                                          |
|------------|----------------------------------------------------------|

> | UMIP Title | Add EURUSD, CHFUSD and GBPUSD as price identifiers       |

Reviewer: Any interest in also proposing the inverses of these?

Author: What would be the point of inverses in our case? @smb2796

Reviewer 2: @smb2796 As mentioned in other UMIPs I don't see the point of adding inverse identifiers and unnecessarily adding to smart contract storage.

| Authors    | Pascal Tallarida (pascal@jarvis.network)                 |
| Status     | Draft                                                    |
| Created    | December 22, 2020                                        |

## Summary
The DVM should support price requests for the EUR/USD, GBP/USD and CHF/USD price index.

## Motivation
The DVM currently does not support the EURUSD, GBPUSD and CHFUSD index.

### Cost: 

> While Forex data are free through open centralized exchange, brokers, and other sources APIs, the most reliable are paid. We propose to use TraderMade's APIs whose pricing is accessible on their website.

Reviewer: Tradermade also has a free tier which should be sufficient for voters

Author: I have added this info

### Opportunity: 
Synthetic tokens that track Forex pairs suche as EURUSD GBPUSD and CHFUSD could be used both for speculationt and hedging, but we see a bigger opportunity for using them as building blocks for helping other DeFi protocols and dApp addressing the european market.

## Technical Specification
The definition of this identifier should be:

> Identifier name: EUR/USD-1e18

Reviewer: This is to indicate that the price identifier is returned as 18 decimals? I would suggest just naming these the same as you have in the `Motivation` section (EURUSD etc)

Reviewer 3: Hey, no it was to say that the jEUR token will have 18 dec.

Author: We want 18 decimals on chain, already added that info in the umip

Base Currency: EUR
Quote Currency: USD

AND

Identifier name: GBP/USD-1e18
Base Currency: GBP
Quote Currency: USD

AND

Identifier name: CHF/USD-1e18
Base Currency: CHF
Quote Currency: USD

Source: https://marketdata.tradermade.com

> Result Processing: Median

Reviewer: Median does not apply as these prices are determined from the results of 1 data source

Author: Fixed

Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.
Price Steps: 0.0001 (4 decimals in more general trading format)

> Rounding: Closest, 0.0001 up

Reviewer: We typically use a slightly convention. Want to start being more explicit with these since there is generally some confusion associated with rounding specifications. My suggestion is:

```suggestion
Rounding: Closest, 0.5 up (> .00005 rounds up, < .00005 rounds down)
```

Author: What is meant here by 0.5 @smb2796. Since a fraction on the Forex market is 0.0001 (depending on the currency pair), so if we are rounding at 0.5 it would cause an increase or decrease of 50%.

Reviewer: It means that a five in the decimal place before your chosen amount of decimal places would round up. Or (> .00005 rounds up, < .00005 rounds down)

Reviewer: Thoughts here?

Author: Got your point and changed it, thanks :)

Decimal: 18 (1e18)
Pricing Interval: 60 seconds
Dispute timestamp rounding: down

## Rationale
Apart from the weekend, there is little to no difference in prices on liquid major Forex pairs like EURUSD, so any price feed could be used; however, for convenience, we recommend using the one of TraderMade.

## Implementation
The value of this identifier for a given timestamp should be determined by querying for the price of EURUSD, GBPUSD and CHFUSD from TraderMade.com for that timestamp. More specifically, users can query “https://marketdata.tradermade.com/api/v1/minute_historical?currency=EURUSD&date_time=2020-11-11-13:01&api_key=apikey” and use the close price as reference. 

Tradermade’s price feed is an aggregated feed from multiple Tier One and Two Banks, Market-Makers and Data Providers. They are popular with Quantitative Traders, Fintech companies and Institutional customers who require a high quality and trusted feed.

> Yet, theur REST API does not work when the Forex market is closed, eem to work on Sunday even though the market opens at 10 pm london time (I asked them). The API just returns the friday OHLC

Reviewer: Can you clarify this sentence. What time do the forex markets close/open etc

Author: Added additional clarifications

> 

Reviewer: I think some clarification on what tokenholders should do when a price request occurs on a weekend would be good. What I'm hearing is that they should use the Friday ohlc, but would be good to make this more explicit.

Author: Added with a later commit :)

As most of the price feed does not provide a price for CHFUSD but USDCHF, the value of this identifier will undergo one additional step: CHFUSD = 1/USDCDF.  Token holders should take care to confirm that the order of the quote and base currency they refer to matches the one being requested by the DVM in the event of a price request.

> Ultimately, how one queries the exchanges should be varied and determined by the voter to ensure that there is no central point of failure.

Reviewer: ```suggestion
TraderMade is provided as an accessible source to query for this data, but ultimately how one queries for these rates should be varied and determined by the voter to ensure that there is no central point of failure.
```

Author: This suggested change is part of a later commit

While it's important for tokenholders to have redundancy in their sources, bots and users that interact with the system in realtime need fast sources of price information. In these cases, it can be assumed that the exchange median is accurate enough

Liquidation and dispute bots should have their own subscription to price feeds. Our price-feed provider’s API documentation can be found [here](https://marketdata.tradermade.com/documentation).

> ## Security considerations

Reviewer: There are also some security considerations around using a singular data source. Are there alternatives that voters could turn to in the case of a TraderMade outage?

Author: Usually voters can turn to basically any price feed out there, since the differences in price feeds between brokers for Forex pairs are almost insignificant.

Reviewer 2: We should still probably officially list more than one data providers in this UMIP so that UMa voters don't unintentionally resolve to a corrupted data providers price.

Adding this new identifier by itself poses a little security risk to the DVM or priceless financial contract users. However, anyone deploying a new priceless token contract referencing this identifier should take care to parameterize the contract appropriately to avoid the loss of funds for synthetic token holders. Additionally, the contract deployer should ensure that there is a network of liquidators and disputers ready to perform the services necessary to keep the contract solvent.

While the price of EURUSD does not vary much across various price feed during the weekday, it may be more challenging to get a unified price feed on the weekend: first, while the underlying market is not closed, most of the broker and banks feeding the APIs price feed are; then, for the one providing a price feed, they often add a very big spread, which makes the price of forex pairs vary significantly by source; a way to mitigate it is to use the last price known. 

> A consequence to this is the existence of gaps between the last quote on Friday to the first one on Sunday; while such gaps rarely exceed +/-2%, which should not put in danger the collateralization ratio, an exceptional event happening the weekend may create a bigger gap and could, in theory, endanger the collateral ratio; yet, it is very unlikely to see +/-20% gap on EURUSD, but it is in theory possible. Nevertheless, due to gaps, a larger number of liquidation and disputes may arise on Sunday 9 pm UTC if the collateral ratio was already getting weaker before the week’s end.

Reviewer: This is well thought out and explained. One additional point I would make is that contracts using this price identifier should have a reasonably high collateralization requirement. This is to combat the danger where a positionholder could be close to the collateralization ratio when entering the weekend, and then a large price gap causes them to be undercollateralized under a 100% ratio. This would mean that their position was no longer profitable to liquidate.

Reviewer 4: I think in the event of a big event (a 20%+ gap on EURUSD) we might see some human judgement, whereby the APIs fail to update but we see liquidators go ahead and liquidate positions.  I suppose that would cause disputes since the API's feeding dispute bots wouldn't know any different, and in that event the tokenholders would be asked to decide on the "market consensus" price -- And I think in the end, we'd arrive at the correct answer.

My tl;dr here is that it _appears_ as if the very very edge case would still have an appropriate (if somewhat messy) outcome.

| UMIP Title | Add EURUSD, CHFUSD and GBPUSD as price identifiers       |
| Authors    | Pascal Tallarida (pascal@jarvis.network)                 |
| Status     | Draft                                                    |
| Created    | December 22, 2020                                        |

## Summary
The DVM should support price requests for the EUR/USD, GBP/USD and CHF/USD price index.

## Motivation
The DVM currently does not support the EURUSD, GBPUSD and CHFUSD index.

### Cost: 
While Forex data are free through open centralized exchange, brokers, and other sources APIs, the most reliable are paid. We propose to use TraderMade's APIs whose pricing is accessible on their website. TraderMade has also a Free Tier which can be used by the voters and would be sufficient to provide the price of a certain asset.

### Opportunity: 
Synthetic tokens that track Forex pairs suche as EURUSD GBPUSD and CHFUSD could be used both for speculationt and hedging, but we see a bigger opportunity for using them as building blocks for helping other DeFi protocols and dApp addressing the european market.

## Technical Specification
The definition of this identifier should be:

> Identifier name: EURUSD

Reviewer 5: This formatting does not conform to the standard. Please make these bullet lists like other UMIPs. See https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-7.md#technical-specification for an example.

Base Currency: EUR
Quote Currency: USD

AND

Identifier name: GBPUSD
Base Currency: GBP
Quote Currency: USD

AND

Identifier name: CHFUSD
Base Currency: CHF
Quote Currency: USD

Source: https://marketdata.tradermade.com

> Result Processing: Mediam (if more than 1 source is utillized), otherwise Exact Price

Reviewer 2: ```suggestion
Result Processing: Median (if more than 1 source is utilized), otherwise Exact Price
```

Input Processing: None. Human intervention in extreme circumstances where the result differs from broad market consensus.
Price Steps: 0.0001 (4 decimals in more general trading format)
Rounding: Closest, 0.0001 up
Decimal: 18 (1e18)
Pricing Interval: 60 seconds
Dispute timestamp rounding: down

## Rationale

> Apart from the weekend, there is little to no difference in prices on liquid major Forex pairs like EURUSD, so any price feed could be used; however, for convenience, we recommend using the one of TraderMade.

Reviewer 2: Using one API source creates a problem for UMA voters if the API source goes down temporarily or does not provide faithful data for a given price request time stamp. I think we should propose at least 2, ideally 3+, data providers. 

It's not about the variance in Fx prices across data providers but about mitigating the chance that a corrupt or incompetent API provider can make resolution more challenging for UMA voters.

Reviewer 5: Additionally, I think that the API used here should match that used in the `CNYUSD` & `USDCNY` UMIPs: https://github.com/UMAprotocol/UMIPs/pull/97/files#r549197519

Ideally, this will enable re-use of the price feeds with two sets of settings without needing custom implementations for each set.

Author: It is not possible to match the API provided in https://github.com/UMAprotocol/UMIPs/pull/97/files#r549197519 since the price feed updates for that API are not in real time. Tradermade however and the other providers I've outlined in my latest commit provide real-time price feed data.

Reviewer: I think Chris's comment was centered around that FX UMIPs should use the same price sources, not that you had to match to the other UMIP.

## Implementation

> The value of this identifier for a given timestamp should be determined by querying for the price of EURUSD, GBPUSD and CHFUSD from TraderMade.com for that timestamp. More specifically, users can query “https://marketdata.tradermade.com/api/v1/minute_historical?currency=EURUSD&date_time=2020-11-11-13:01&api_key=apikey” and use the close price as reference. 

Reviewer 5: Is there an option for polling this API that does not require an API key?

Tradermade’s price feed is an aggregated feed from multiple Tier One and Two Banks, Market-Makers and Data Providers. They are popular with Quantitative Traders, Fintech companies and Institutional customers who require a high quality and trusted feed.

Over the weekend the REST API does not return any price, but we can request the price of a certain moment before the market close (as ex: the closing price of Friday).

> Due to unavailability of price feed for any financial asset over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which esentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset was created, if it was created over the weekend. If not the closing price on Friday for a certain asset should be a navigating point in calculating the collateralization ratio of a position and in the liquidation process.

Reviewer 2: So just to be clear, we're establishing a convention that if a request timestamp takes place on a weekend (or any day that the Fx market is closed), then we should use the last traded tick as the price? 

I think this is reasonable, but we should make this more clear in the UMIP itself. Perhaps in the "technical specification" section.

> Forex markets are usually open throughout bussiness days from 10:00 PM UTC time on Sunday, until 9:00 PM UTC time on Friday. However it is good to note that some brokers might close their price feeds 5-10 minutes prior to the official market closing time, thus the latest price given from the broker will be used as a official price for the asset. We should also note here that during those minutes the price remains flat without any changes, however theoretically it is possible to have a stronger price movement.

Reviewer 2: ```suggestion
Forex markets are usually open throughout business days from 10:00 PM UTC time on Sunday, until 9:00 PM UTC time on Friday. However it is good to note that some brokers might close their price feeds 5-10 minutes prior to the official market closing time, thus the latest price given from the broker will be used as a official price for the asset. We should also note here that during those minutes the price remains flat without any changes, however theoretically it is possible to have a stronger price movement.
```

> As most of the price feed does not provide a price for CHFUSD but USDCHF, the value of this identifier will undergo one additional step: CHFUSD = 1/USDCDF.  Token holders should take care to confirm that the order of the quote and base currency they refer to matches the one being requested by the DVM in the event of a price request.

Reviewer 2: Why don't we just add a USDCHF identifier instead of a CHFUSD to reduce steps for voters?

Author: @nicholaspai USDCHF price feed is not supported by price feed providers, since the original display of the currency pair is CHFUSD, thus voters will have to make the switch calculation.

Reviewer 2: What do you mean by "As most of the price feed does not provide a price for CHFUSD but USDCHF, the value of this identifier will undergo one additional step: CHFUSD = 1/USDCDF" then?

Let's just use the identifier that requires fewer calculation steps for the voters, so if price feed providers only support CHFUSD then let's do that. Am I understanding correctly?

Author: @nicholaspai Sorry my previous comment was mistaken. However since in the derivative we use CHF/USD it is better to have CHF/USD as price identifier, however as mentioned in the umip voters will have to get USD/CHF price feed and calculate the reversed price.

A consequence to this is the existence of gaps between the last quote on Friday to the first one on Sunday; while such gaps rarely exceed +/-2%, which should not put in danger the collateralization ratio, an exceptional event happening the weekend may create a bigger gap and could, in theory, endanger the collateral ratio; yet, it is very unlikely to see +/-20% gap on EURUSD, but it is in theory possible. Nevertheless, due to gaps, a larger number of liquidation and disputes may arise on Sunday 9 pm UTC if the collateral ratio was already getting weaker before the week’s end. In order to mitigate this theoretical risk a high collateralization requirement (120%+) should be set or a reserve fund mechanism should be implemented in order to automatically rebalance the collateralization ratio above 100% (in case it falls below 100%), thus making the liquidation process profitable again.

> In the past the CHF was pegged to the EUR, but in 2015 this pegged was removed by the Swiss Central Bank in a surprise move, which caused a huge volatile movement on the CHF price. Such movement could put the CHF below the 100% collateralization ratio, although it is highly unlikely to happen. This risk could appear again one day if the peg is reinstated. In the forementioned situation the CHF lost 20% within one day. However with a higher required collateralization ratio such events can be mitigated.

Reviewer 2: Good color thanks!

Tradermade’s price feed is an aggregated feed from multiple Tier One and Two Banks, Market-Makers and Data Providers. They are popular with Quantitative Traders, Fintech companies and Institutional customers who require a high quality and trusted feed.

### Weekend timestamp

Over the weekend the REST API does not return any price, but we can request the price of a certain moment before the market close (as ex: the closing price of Friday).

Due to unavailability of price feed for any financial asset over the weekend, tokenholders and users will be using the latest known price, before the market close on Friday, which esentially is the closing price of the asset from Friday. Same goes in a case of a liquidation process - the liquidator should use the last price from Friday in order to match with the price on which a synthetic asset was created, if it was created over the weekend. If not the closing price on Friday for a certain asset should be a navigating point in calculating the collateralization ratio of a position and in the liquidation process.

> If a request timestamp takes place on a weekend or any other day the Forex market is closed, voters should use the latest tick as the price. For the weekend that would be the closing price of the asset on Friday.

Reviewer: Thank you for this clarification. Seems sufficient to me

TraderMade is provided as an accessible source to query for this data, but ultimately how one queries for these rates should be varied and determined by the voter to ensure that there is no central point of failure.
While it's important for tokenholders to have redundancy in their sources, bots and users that interact with the system in realtime need fast sources of price information. In these cases, it can be assumed that the exchange median is accurate enough

### Price feed - liquidations and disputes

Liquidation and dispute bots should have their own subscription to price feeds. Our price-feed provider’s API documentation can be found [here](https://marketdata.tradermade.com/documentation).

In the case of a TraderMade outage voters can turn to any other available price feed API or a broker API, as the price feeds for the forementioned financial assets does not differ much between different providers. There might be some slight differences, however they are quite insignificant and would not affect the liquidation or dispute processes. In that case we'll use a median price from the different price feeds provided by the voters.

## Additional price feed providers

> ### FXMarketAPI

Reviewer: Thanks for adding this. Great alternative for voters
